rfd/0202-db-multi-session-mfa.md

```diff
@@ -0,0 +1,309 @@
+https://github.com/gravitational/teleport/blob/6cfeacb547d868be23c0320d1d000449f7c7960c/lib/auth/grpcserver.go#L2028-L2050
+
+The downgrade will be lifted in next major release version.
+
+Newer clients will not be able to set the new role option or request the new
+scope. They will continue use per-session MFA with older servers.
+
+### The `tsh db exec` command
+
+General flow of the command:
+- Fetch databases (either specified directly or through search).
+- For each database:
+  - Prompt MFA:
+    - If `per-session` MFA is required.
+    - Or, if first time requesting MFA for `multi-session` or shared `multi-session`
+      response is expired.
+    - No MFA if not required.
+  - Starts a local proxy in tunnel mode for this database (regardless of cluster
+    proxy listener mode).
+  - Craft a command for `os.exec`. Outputs are printed to `stdout` unless
+    `--log-dir` is specified.
+  - Execute the command.
+ 
+The command supports searching database by specifying one the following flags:
+- `--search`: List of comma separated search keywords or phrases enclosed in
+  quotations, e.g. `--search=foo,bar`.
+- `--labels`: List of comma separated labels to filter by labels, e.g.
+  `key1=value1,key2=value2`.
+- `--query` will NOT be supported for this command as it is harder to use than
+  the other options and the name `--query` may be confused with the database
+  query that needs to be executed.
+The command presents the search results then asks user to confirm before
+proceeding. `--skip-confirm` can be used to skip the confirmation.
+
+If no search parameters are used, a list of database names can be specified
+through `--dbs db1,db2,...`.
+
+For the MVP implementation, only PostgreSQL and MySQL databases will be
+supported. And a warning will be printed if the target databases have different
+protocols (e.g. `postgres` vs `mysql`).
+
+For databases that require per-session MFA, a prompt will still be presented
+per database.
+ 
+#### Possible future enhancements for `tsh db exec`
+- `tsh db exec --exec-config` to support a config file which allows specifying
+  different flags like `--db-user`, `--db-name`, `--exec-query` per target
+  database or per search.
+- `tsh db exec --command-template` to support custom command template like `tsh
+  db exec --exec-command "bash -c './my_script.sh {{.DB_SERVICE}} {{.DB_USER}}
+  {{.DB_NAME}} {{.DB_LOCAL_PORT}}'"`. An env var `TSH_UNSTABLE_DB_EXEC_COMMAND`
+  can be supported for the initial MVP.
+
+### Max connections
+
+The new `tsh db exec` command with the `--max-connections` flag makes it easier
+to for the client to flood the backend.
+
+A new role option is added to set a maximum number of in-flight connections:
+```diff
+kind: role
+version: v7
+spec:
+  options:
+    # limit number of concurrent Kubernetes sessions per user
+    max_kubernetes_connections: 1
++   # limit number of concurrent database sessions per user
++   max_db_connections: 5
```

Comment:
Hrm, I assume this new setting will take effect for "regular" connections as well, e.g. if I just do "tsh db connect" 5 times? I imagine opening multiple sessions for the same database is pretty common use-case, what would the default value be? And what would be the behavior if you have more than the limit?

I agree it would be good to have it but I think we should leave it for phase 2 maybe, seems like it's a new feature that's tangentially related to multi-MFA. TBH for now I would rather have that "tsh db exec" is smart enough not to DDOS the cluster itself with max connections in flight.